edx_django_utils/monitoring/middleware.py

```diff
@@ -0,0 +1,186 @@
+"""
+Middleware for handling the storage, aggregation, and reporting of custom
+metrics for monitoring.
+
+At this time, the custom metrics can only be reported to New Relic.
+
+This middleware will only call on the newrelic agent if there are any metrics
+to report for this request, so it will not incur any processing overhead for
+request handlers which do not record custom metrics.
+
+"""
+import logging
+from uuid import uuid4
+
+import psutil
+import waffle
+from django.utils import six
+from edx_django_utils.cache import RequestCache
+
+log = logging.getLogger(__name__)
+try:
+    import newrelic.agent
+except ImportError:
+    log.warning("Unable to load NewRelic agent module")
+    newrelic = None  # pylint: disable=invalid-name
+
+
+_DEFAULT_NAMESPACE = 'edx_django_utils.monitoring'
+_REQUEST_CACHE_NAMESPACE = '{}.custom_metrics'.format(_DEFAULT_NAMESPACE)
+
+
+class MonitoringCustomMetrics(object):
```

Comment:
These Middleware have a dependency on the RequestCache, and thus the RequestCacheMiddleware.  This should be clarified in the docs, and possibly protected against as described in this comment: https://github.com/edx/edx-platform/pull/18725#discussion_r209310957 . It should either log an error, or throw an exception if it can do the same that Django does for these types of issues.